File: io/java/src/main/java/org/shypl/biser/io/Entity.java
package org.shypl.biser.io;

import org.shypl.common.util.StringUtils;

import java.util.LinkedHashMap;
import java.util.Map;

public abstract class Entity {
	@Override
	public String toString() {
		Map<String, String> fields = new LinkedHashMap<>();
		_toString(fields);
		return StringUtils.toString(fields);
	}
	
	public int _id() {
		return 0;
	}
	
	protected void _encode(DataWriter writer) {
	}
	
	protected void _decode(DataReader reader) {
	}
	
	protected void _toString(Map<String, String> fields) {
	}
}

File: kotlin/biser-kotlin-js/src/main/kotlin/org/shypl/biser/io/Entity.kt
package org.shypl.biser.io

import org.shypl.biser.StringUtils

abstract class Entity {
	override fun toString(): String {
		val fields = LinkedHashMap<String, String>()
		_toString(fields)
		return StringUtils.toString(fields)
	}
	
	open fun _id(): Int {
		return 0
	}
	
	open fun _encode(writer: DataWriter) {}
	
	open fun _decode(reader: DataReader) {}
	
	protected open fun _toString(fields: MutableMap<String, String>) {}
}